File: python-service/src/request_validator.py
import re
from datetime import datetime
from threading import Lock
from typing import List, Dict, Optional

MAX_CONTENT_SIZE = 1_000_000
MAX_PATH_LENGTH = 500
ALLOWED_LANGUAGES = ['go', 'python', 'ruby', 'javascript', 'typescript', 'java']

validation_errors: List[Dict] = []
validation_lock = Lock()


class ValidationError:
    def __init__(self, field: str, reason: str):
        self.field = field
        self.reason = reason
        self.timestamp = datetime.now().isoformat()

    def to_dict(self) -> Dict:
        return {
            'field': self.field,
            'reason': self.reason,
            'timestamp': self.timestamp
        }


def validate_review_request(data: Dict) -> List[ValidationError]:
    errors = []

    content = data.get('content')
    language = data.get('language')

    if not content:
        errors.append(ValidationError('content', 'Content is required and cannot be empty'))
    elif len(content) > MAX_CONTENT_SIZE:
        errors.append(ValidationError('content', f'Content exceeds maximum size of {MAX_CONTENT_SIZE} bytes'))
    elif contains_null_bytes(content):
        errors.append(ValidationError('content', 'Content contains invalid null bytes'))

    if language and language not in ALLOWED_LANGUAGES:
        errors.append(ValidationError('language', f'Language must be one of: {", ".join(ALLOWED_LANGUAGES)}'))

    log_validation_errors(errors)
    return errors


def validate_statistics_request(data: Dict) -> List[ValidationError]:
    errors = []

    files = data.get('files')

    if not files:
        errors.append(ValidationError('files', 'Files array is required'))
    elif not isinstance(files, list):
        errors.append(ValidationError('files', 'Files must be an array'))
    elif len(files) == 0:
        errors.append(ValidationError('files', 'Files array cannot be empty'))
    elif len(files) > 1000:
        errors.append(ValidationError('files', 'Files array cannot exceed 1000 entries'))

    log_validation_errors(errors)
    return errors


def sanitize_input(input_str: Optional[str]) -> Optional[str]:
    if input_str is None:
        return None
    if not isinstance(input_str, str):
        return str(input_str)

    sanitized = []
    for char in input_str:
        code = ord(char)
        if code == 0 or (1 <= code <= 8) or code == 11 or code == 12 or (14 <= code <= 31) or code == 127:
            continue
        if char in ['\n', '\r', '\t']:
            sanitized.append(char)
        elif not char.isprintable() and char not in ['\n', '\r', '\t']:
            continue
        else:
            sanitized.append(char)

    return ''.join(sanitized)


def sanitize_request_data(data: Dict) -> Dict:
    sanitized = data.copy()

    if 'content' in sanitized and isinstance(sanitized['content'], str):
        sanitized['content'] = sanitize_input(sanitized['content'])

    if 'language' in sanitized and isinstance(sanitized['language'], str):
        sanitized['language'] = sanitize_input(sanitized['language'])

    if 'path' in sanitized and isinstance(sanitized['path'], str):
        sanitized['path'] = sanitize_input(sanitized['path'])

    return sanitized


def contains_null_bytes(content: str) -> bool:
    return '\x00' in content


def contains_path_traversal(path: str) -> bool:
    return '..' in path or '~/' in path


def log_validation_errors(errors: List[ValidationError]):
    if not errors:
        return

    with validation_lock:
        validation_errors.extend([error.to_dict() for error in errors])
        keep_recent_errors()


def keep_recent_errors():
    global validation_errors
    if len(validation_errors) > 100:
        validation_errors = validation_errors[-100:]


def get_validation_errors() -> List[Dict]:
    with validation_lock:
        return validation_errors.copy()


def clear_validation_errors():
    with validation_lock:
        validation_errors.clear()

File: ruby-service/lib/request_validator.rb
require 'json'

class RequestValidator
  VALIDATION_ERRORS = []
  MAX_CONTENT_SIZE = 1_000_000
  MAX_PATH_LENGTH = 500
  ALLOWED_LANGUAGES = %w[go python ruby javascript typescript java].freeze

  class ValidationError < StandardError
    attr_reader :field, :reason

    def initialize(field, reason)
      @field = field
      @reason = reason
      super("Validation failed for #{field}: #{reason}")
    end

    def to_hash
      { field: @field, reason: @reason }
    end
  end

  def self.validate_analyze_request(params)
    errors = []

    content = params['content'] || params[:content]
    path = params['path'] || params[:path]

    if content.nil? || content.empty?
      errors << ValidationError.new('content', 'Content is required and cannot be empty')
    elsif content.length > MAX_CONTENT_SIZE
      errors << ValidationError.new('content', "Content exceeds maximum size of #{MAX_CONTENT_SIZE} bytes")
    elsif contains_null_bytes?(content)
      errors << ValidationError.new('content', 'Content contains invalid null bytes')
    end

    if path && path.length > MAX_PATH_LENGTH
      errors << ValidationError.new('path', "Path exceeds maximum length of #{MAX_PATH_LENGTH} characters")
    end

    if path && contains_path_traversal?(path)
      errors << ValidationError.new('path', 'Path contains potential directory traversal')
    end

    log_validation_errors(errors) unless errors.empty?

    errors
  end

  def self.validate_diff_request(params)
    errors = []

    old_content = params['old_content'] || params[:old_content]
    new_content = params['new_content'] || params[:new_content]

    if old_content.nil? || old_content.empty?
      errors << ValidationError.new('old_content', 'Old content is required')
    elsif old_content.length > MAX_CONTENT_SIZE
      errors << ValidationError.new('old_content', "Old content exceeds maximum size")
    end

    if new_content.nil? || new_content.empty?
      errors << ValidationError.new('new_content', 'New content is required')
    elsif new_content.length > MAX_CONTENT_SIZE
      errors << ValidationError.new('new_content', "New content exceeds maximum size")
    end

    log_validation_errors(errors) unless errors.empty?

    errors
  end

  def self.validate_metrics_request(params)
    errors = []

    content = params['content'] || params[:content]

    if content.nil? || content.empty?
      errors << ValidationError.new('content', 'Content is required for metrics')
    elsif content.length > MAX_CONTENT_SIZE
      errors << ValidationError.new('content', "Content exceeds maximum size")
    end

    log_validation_errors(errors) unless errors.empty?

    errors
  end

  def self.sanitize_input(input)
    return nil if input.nil?
    return input unless input.is_a?(String)

    sanitized = input.gsub(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/, '')
    sanitized = sanitized.force_encoding('UTF-8')
    sanitized.valid_encoding? ? sanitized : sanitized.encode('UTF-8', invalid: :replace, undef: :replace)
  end

  def self.get_validation_errors
    VALIDATION_ERRORS.dup
  end

  def self.clear_validation_errors
    VALIDATION_ERRORS.clear
  end

  private

  def self.contains_null_bytes?(content)
    content.include?("\x00")
  end

  def self.contains_path_traversal?(path)
    path.include?('..') || path.include?('~/')
  end

  def self.log_validation_errors(errors)
    timestamp = Time.now.iso8601
    errors.each do |error|
      VALIDATION_ERRORS << {
        timestamp: timestamp,
        field: error.field,
        reason: error.reason
      }
    end
    keep_recent_errors
  end

  def self.keep_recent_errors
    VALIDATION_ERRORS.shift while VALIDATION_ERRORS.length > 100
  end
end
